pkg/statsd/statsd.go

```diff
@@ -145,22 +166,22 @@ func (s *Server) RunWithCustomSocket(ctx context.Context, sf SocketFactory) erro
 	}
 
 	// 8. Start the Receiver
-	// Open socket
-	c, err := sf()
-	if err != nil {
-		return err
-	}
-	defer func() {
-		// This makes receivers error out and stop
-		if e := c.Close(); e != nil {
-			log.Warnf("Error closing socket: %v", e)
-		}
-	}()
-
 	receiver := NewDatagramReceiver(datagrams, s.ReceiveBatchSize, statser)
 	stage = stgr.NextStage()
 	stage.StartWithContext(receiver.RunMetrics)
 	for r := 0; r < s.MaxReaders; r++ {
+		// Open socket
+		c, err := sf()
+		if err != nil {
+			return err
+		}
+		defer func(c net.PacketConn) {
+			// This makes receivers error out and stop
+			if e := c.Close(); e != nil {
```

Comment:
Shouldn't be a problem, but in a non-shared case, this will close the same socket multiple times?

Reply:
The factory function could return a `func() error` that should be used to free up the resources. And the implementation could call `Close()` just one time for non-shared case.

Reply:
I overrode `Close()` and wrapped it in a sync.Once

Reply:
:meh: it doesn't panic.